controllers/cloudstackmachine_controller.go

```diff
@@ -21,265 +21,227 @@ import (
-			if zone := csCluster.Status.Zones.GetByName(csMachine.Spec.ZoneName); zone != nil { // ZoneName Specified.
-				csMachine.Status.ZoneID = zone.ID
+		} else if r.ReconciliationSubject.Spec.ZoneName != "" {
+			if zone := r.CSCluster.Status.Zones.GetByName(r.ReconciliationSubject.Spec.ZoneName); zone != nil { // ZoneName Specified.
+				r.ReconciliationSubject.Status.ZoneID = zone.ID
 			} else {
-				return ctrl.Result{}, errors.Errorf("could not find zone by zoneName: %s", csMachine.Spec.ZoneName)
+				return ctrl.Result{}, errors.Errorf("could not find zone by zoneName: %s", r.ReconciliationSubject.Spec.ZoneName)
 			}
 		} else { // No Zone Specified, pick a Random Zone.
-			zones := make([]string, len(csCluster.Status.Zones))
-			zidx := 0
-			for zoneID := range csCluster.Status.Zones {
-				zones[zidx] = zoneID
-				zidx++
-			}
-			randNum := (rand.Int() % len(csCluster.Spec.Zones)) // #nosec G404 -- weak crypt rand doesn't matter here.
-			csMachine.Status.ZoneID = zones[randNum]
+			randNum := (rand.Int() % len(r.Zones.Items)) // #nosec G404 -- weak crypt rand doesn't matter here.
+			r.ReconciliationSubject.Status.ZoneID = r.Zones.Items[randNum].Spec.ID
+		}
+	}
+	for idx, zone := range r.Zones.Items {
+		if zone.Spec.ID == r.ReconciliationSubject.Status.ZoneID {
+			r.FailureDomain = &r.Zones.Items[idx]
+			break
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// GetOrCreateVMInstance creates or gets a VM instance.
+// Implicitly it also fetches its bootstrap secret in order to create said instance.
+func (r *CloudStackMachineReconciliationRunner) GetOrCreateVMInstance() (retRes ctrl.Result, reterr error) {
+	if r.CAPIMachine.Spec.Bootstrap.DataSecretName == nil {
+		return r.RequeueWithMessage("Bootstrap DataSecretName not yet available.")
+	}
+	r.Log.Info("Got Bootstrap DataSecretName.")
+
+	// Get the CloudStackZone for the Machine.
+	var machineZone infrav1.CloudStackZone
+	for _, zone := range r.Zones.Items {
+		machineZone = zone
+		if zone.Spec.ID == r.ReconciliationSubject.Status.ZoneID {
+			break
 		}
 	}
 
+	// Get the kubeadm bootstrap secret for this machine.
 	secret := &corev1.Secret{}
-	key := types.NamespacedName{Namespace: capiMachine.Namespace, Name: *capiMachine.Spec.Bootstrap.DataSecretName}
+	key := types.NamespacedName{Namespace: r.CAPIMachine.Namespace, Name: *r.CAPIMachine.Spec.Bootstrap.DataSecretName}
 	if err := r.Client.Get(context.TODO(), key, secret); err != nil {
 		return ctrl.Result{}, err
 	}
-
-	// Create VM (or Fetch if present).
 	if data, isPresent := secret.Data["value"]; isPresent {
-		if err := r.CS.GetOrCreateVMInstance(csMachine, capiMachine, csCluster, string(data)); err == nil {
-			if !controllerutil.ContainsFinalizer(csMachine, infrav1.MachineFinalizer) { // Fetched or Created?
-				log.Info("CloudStack instance Created", "instanceStatus", csMachine.Status)
-				controllerutil.AddFinalizer(csMachine, infrav1.MachineFinalizer)
+		if err := r.CS.GetOrCreateVMInstance(
```

Comment:
More nesting than I think is desirable.  I suggest keeping the "secret not ready yet" return close to the test for secret readiness, and have the GetOrCreateVM at the outer scope, since it's a primary workflow step.  I've been seeing this called out in some style guides.

Reply:
Yeah, that one was kinda ugly. Will fix.